modules/tensor_mechanics/src/userobjects/nodal_patch_recovery/NPRBase.C

```diff
@@ -0,0 +1,173 @@
+                                [](const InputParameters &, InputParameters & rm_params) {
+                                  rm_params.set<unsigned short>("layers") = 2;
+                                });
+
+  params.addParamNamesToGroup("patch_polynomial_order", "Advanced");
+
+  return params;
+}
+
+NPRBase::NPRBase(const InputParameters & parameters)
+  : ElementUserObject(parameters),
+    _qp(0),
+    _patch_polynomial_order(
+        static_cast<unsigned int>(getParam<MooseEnum>("patch_polynomial_order"))),
+    _multi_index(MathUtils::multiIndex(_mesh.dimension(), _patch_polynomial_order)),
+    _q(_multi_index.size())
+{
+}
+
+Real
+NPRBase::nodalPatchRecovery(const Point & x, const std::vector<dof_id_type> & elem_ids) const
+{
+  // before we go, check if we have enough sample points for solving the least square fitting
+  if (_q_point.size() * elem_ids.size() < _q)
+    mooseError("There are not enough sample points to recover the nodal value, try reducing the "
+               "polynomial order or using a higher-order quadrature scheme.");
+
+  // Assemble the least squares problem over the patch
+  RealEigenMatrix A = RealEigenMatrix::Zero(_q, _q);
+  RealEigenVector b = RealEigenVector::Zero(_q);
+  for (auto elem_id : elem_ids)
+  {
+    A += _Ae.at(elem_id);
+    b += _be.at(elem_id);
+  }
+
+  // Solve the least squares fitting
+  RealEigenVector coef = A.completeOrthogonalDecomposition().solve(b);
+
+  // Compute the fitted nodal value
+  RealEigenVector p = evaluateBasisFunctions(x);
+  return p.dot(coef);
+}
+
+RealEigenVector
+NPRBase::evaluateBasisFunctions(const Point & q_point) const
+{
+  RealEigenVector p(_q);
+  Real polynomial;
+  for (unsigned int r = 0; r < _multi_index.size(); r++)
+  {
+    polynomial = 1.0;
+    for (unsigned int c = 0; c < _multi_index[0].size(); c++)
+      for (unsigned int p = 0; p < _multi_index[r][c]; p++)
+        polynomial *= q_point(c);
+    p(r) = polynomial;
+  }
+  return p;
+}
+
+void
+NPRBase::initialize()
+{
+  _Ae.clear();
+  _be.clear();
+}
+
+void
+NPRBase::execute()
+{
+  RealEigenMatrix Ae = RealEigenMatrix::Zero(_q, _q);
+  RealEigenVector be = RealEigenVector::Zero(_q);
+  for (_qp = 0; _qp < _qrule->n_points(); _qp++)
+  {
+    RealEigenVector p = evaluateBasisFunctions(_q_point[_qp]);
+    Ae += p * p.transpose();
+    be += computeValue() * p;
+  }
+
+  dof_id_type elem_id = _current_elem->id();
+  _Ae[elem_id] = Ae;
+  _be[elem_id] = be;
+}
+
+void
+NPRBase::finalize()
+{
+  // When calling nodalPatchRecovery, we may need to know _Ae and _be on algebraically ghosted
+  // elements. However, this userobject is only run on local elements, so we need to querry those
+  // information from other processors in this finalize() method.;
+
+  // Populate algebraically elements to querry
+  std::unordered_map<processor_id_type, std::vector<dof_id_type>> querry_ids;
+  const ConstElemRange evaluable_elem_range = _fe_problem.getEvaluableElementRange();
```

Comment:
This is asking for evaluable elements based on the DofMap in nonlinear system. Ideally, this should be the intersection of {evaluable elements from nonlinear system} and {evaluable elements from auxiliary system}. We may need to wait for the next libmesh update where @lindsayad will add a predicate for that purpose.

Reply:
We will have to wait. I've written the code. Just need to do some testing and submit the PR

Reply:
This: https://github.com/libMesh/libmesh/pull/3005 is the one, right? @hugary1995 will you be able to work on this anytime soon, or should we try and fix this up for you?

Reply:
Yes, that's the one. I'm still waiting for a libmesh update #18831. Once it's merged I can work on this.

Reply:
Oh wait... Maybe it's already merged.

Reply:
Okay, I fixed it. This is ready for your review now.